### PR TITLE
Fix-Transaction-GTM-Event

### PR DESCRIPTION
**Pull request by TheNoodleMoose**
Clark realized that transaction events aren't actually firing on give stage or prod. I think it has something to do with where the call is made, and the dataLayer being uninitialized during page loads(from review checkout to the thank you summary page). So over the weekend, I thought of a solution to prevent duplicate transaction event calls on the thank you summary page. Basically when the purchase is made I store their cart in localStorage in order for use in the transaction event. I then check localStorage for ```transactionId``` to see if it matches the purchase id coming into the function. If it doesn't I know this is a new purchase, so we need to send a transaction event and also store this purchase id in localStorage as ```transactionId```. That way if they refresh the page or something to that extent, the ```transactionId``` is set in localStorage, and the id's will now match so it won't send a duplicate transaction event. I then also remove the cart from local storage since it's no longer needed. Let me know what you think of this solution. I thought going with localStorage would be better anyway since the digitalData layer will probably go away sometime in the future.

### CODE REVIEW

File: src/app/checkout/step-3/step-3.component.js
Comment (TheNoodleMoose) [inline src/app/checkout/step-3/step-3.component.js:177]: This solution also gets rid of the additional calls to the api which is great.

File: src/app/analytics/analytics.factory.js
Comment (wrandall22) [inline src/app/analytics/analytics.factory.js:492]: This `const` should only be alive within this `if` block being defined like this. Shouldn't it be defined above the `if` block instead?

Comment (wrandall22) [reply src/app/analytics/analytics.factory.js:492]: Nevermind, it is also used inside the `if`. Hard to see in GitHub.

File: src/app/analytics/analytics.factory.js
Comment (wrandall22) [inline src/app/analytics/analytics.factory.js:491]: So, it looks like `transactionId` is never cleared from local storage. I wonder if this could introduce a security hole, especially on public computers. Maybe it should be cleared on sign-in and/or sign-out events?

Comment (TheNoodleMoose) [reply src/app/analytics/analytics.factory.js:491]: I also thought about doing sessionStorage instead, do you think that would be better?

Comment (wrandall22) [reply src/app/analytics/analytics.factory.js:491]: How would the thank you summary load in a new browser session?

Comment (TheNoodleMoose) [reply src/app/analytics/analytics.factory.js:491]: You can disregard that last part haha. Should I go with sessionStorage instead so we make sure it's cleared after their session ends regardless of being signed in or not?

Comment (TheNoodleMoose) [reply src/app/analytics/analytics.factory.js:491]: Actually, I think that is the way to go because we do the same thing for the ```lastPurchaseLink```, ```storedCvvs```, etc

Comment (wrandall22) [reply src/app/analytics/analytics.factory.js:491]: So, I thought of one way that might hit the thank you page without going through checkout, which is site crawlers. You can go to https://give-stage2.cru.org/thank-you.html directly without a cart. It throws an error, but we should probably also account for that case of bots/web crawlers.

File: src/app/analytics/analytics.factory.js
Comment (wrandall22) [inline src/app/analytics/analytics.factory.js:485]: `sessionStorage` is not currently available to this factory I think.

Comment (TheNoodleMoose) [reply src/app/analytics/analytics.factory.js:485]: It seems to work. Is it because ```$window``` is exposed in the factory? The other places we use sessionStorage just do ```this.sessionStorage = $window.sessionStorage```
Screen Shot 2020-11-23 at 10 01 01 AM

Comment (wrandall22) [reply src/app/analytics/analytics.factory.js:485]: Interesting. Not sure why.

File: src/app/analytics/analytics.factory.js
Comment (wrandall22) [inline src/app/analytics/analytics.factory.js:487]: Perhaps in order to account for web crawlers, we should ensure `purchaseData` is defined?

Comment (TheNoodleMoose) [reply src/app/analytics/analytics.factory.js:487]: Yep, sounds good to me!

**Comment by wrandall22** [comment]
Feel free to push to staging/test in stage.

**Comment by TheNoodleMoose** [comment]
@wrandall22 Clark gave the thumbs up that this works as expected now on staging. Should I wait for a code review from Scotty or do you think this is good as is for prod?

**Comment by wrandall22** [comment]
I'd like to hear @OzzieOrca opinion on the above conversation.

File: src/app/analytics/analytics.factory.js
Comment (OzzieOrca) [reply src/app/analytics/analytics.factory.js:485]: `localStorage` and `sessionStorage` are both JS globals because everything on the `window` object is considered to be global. Best practices in AngularJS seem to suggest you get these from `$window`. I think that's to make mocking in tests a little easier. But idk if it matters too much.

**Review by wrandall22 — APPROVED** [review]

**Pull request merged**
